examples/auction/src/lib.rs

```diff
@@ -3,206 +3,191 @@ use core::fmt::Debug;
+    host: &impl HasHost<State, StateApiType = S>,
 ) -> ReceiveResult<ViewableState> {
     Ok(host.state().viewable_state.clone())
 }
 
-/// Receive function used to finalize the auction, returning all bids to their
-/// senders, except for the winning bid
+/// Receive function used to finalize the auction. It sends the highest bid (the
+/// current balance of this smart contract) to the owner of this smart contract.
 #[receive(contract = "auction", name = "finalize", mutable)]
 fn auction_finalize<S: HasStateApi>(
     ctx: &impl HasReceiveContext,
-    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    host: &mut impl HasHost<State, StateApiType = S>,
 ) -> Result<(), FinalizeError> {
     let state = host.state();
+    // Ensure that auction has not been finalized yet
     ensure!(
         state.viewable_state.auction_state == AuctionState::NotSoldYet,
         FinalizeError::AuctionFinalized
     );
 
     let slot_time = ctx.metadata().slot_time();
-    ensure!(slot_time > state.viewable_state.expiry, FinalizeError::AuctionStillActive);
-
-    let owner = ctx.owner();
-
-    let balance = host.self_balance();
-    if balance == Amount::zero() {
-        Ok(())
-    } else {
-        if host.invoke_transfer(&owner, state.viewable_state.highest_bid).is_err() {
-            bail!(FinalizeError::BidMapError);
-        }
-        let mut remaining_bid = None;
-        // Return bids that are smaller than highest
-        for (addr, amnt) in state.bids.iter() {
-            if *amnt < state.viewable_state.highest_bid {
-                if host.invoke_transfer(&*addr, *amnt).is_err() {
-                    bail!(FinalizeError::BidMapError);
-                }
-            } else {
-                ensure!(remaining_bid.is_none(), FinalizeError::BidMapError);
-                remaining_bid = Some((addr, amnt));
-            }
-        }
-        // Ensure that the only bidder left in the map is the one with the highest bid
-        match remaining_bid {
-            Some((addr, amount)) => {
-                ensure!(*amount == state.viewable_state.highest_bid, FinalizeError::BidMapError);
-                host.state_mut().viewable_state.auction_state = AuctionState::Sold(*addr);
-                Ok(())
-            }
-            None => bail!(FinalizeError::BidMapError),
-        }
+    // Ensure that auction has ended already
+    ensure!(slot_time > state.viewable_state.end, FinalizeError::AuctionStillActive);
+
+    if let Some(account_address) = state.viewable_state.highest_bidder {
+        // Marking the highest bid (the last bidder) as winner of the auction
+        host.state_mut().viewable_state.auction_state = AuctionState::Sold(account_address);
+        let owner = ctx.owner();
+        let balance = host.self_balance();
+        // Sending the highest bid (the balance of this contract) to the owner of this
+        // contract
+        if host.invoke_transfer(&owner, balance).is_err() {}
```

Comment:
Again, add some comment to why not to deal reject in case of an error.
I would probably have rejected the contract in this case.